src/main/java/sirius/kernel/commons/OptimisticLockAlgorithm.java

```diff
@@ -0,0 +1,116 @@
+ *         hard to take back.
+ *     </li>
+ * </ol>
+ * <p>
+ * If these preconditions are fullfilled, optimistic locking can provide an extremely efficient way
+ * of handling critical sections (tasks which might run in parallel) as there is almost no overhead.
+ *
+ * @param <R> the type of results being produced by the algorithm
+ */
+public class OptimisticLockAlgorithm<R> {
+
+    private int minWaitMillis = 10;
+    private int maxWaitMillis = 250;
+    private int numberOfRetries = 5;
+
+    /**
+     * Specifies the maximal number of retries bevore giving up.
+     *
+     * @param numberOfRetries the maximal number of attempts to perform and validate the action
+     * @return the helper itself for fluent method calls
+     */
+    public OptimisticLockAlgorithm<R> withRetries(int numberOfRetries) {
+        this.numberOfRetries = numberOfRetries;
+        return this;
+    }
+
+    /**
+     * Specifies the minimal wait time before retrying.
+     * <p>
+     * If the algorithm detects a collision, it performs a cleanup and then waits a random
+     * period of time, so that two threads or nodes don't constantly collide. This specifies the
+     * lower bound of this period.
+     *
+     * @param minWaitMillis the minimal wait time in milliseconds
+     * @return the helper itself for fluent method calls
+     */
+    public OptimisticLockAlgorithm<R> withMinWaitMillis(int minWaitMillis) {
+        this.minWaitMillis = minWaitMillis;
+        return this;
+    }
+
+    /**
+     * Specifies the maximal wait time before retrying.
+     * <p>
+     * If the algorithm detects a collision, it performs a cleanup and then waits a random
+     * period of time, so that two threads or nodes don't constantly collide. This specifies the
+     * upper bound of this period.
+     *
+     * @param maxWaitMillis the maximal wait time in milliseconds
+     * @return the helper itself for fluent method calls
+     */
+    public OptimisticLockAlgorithm<R> withMaxWaitMillis(int maxWaitMillis) {
+        this.maxWaitMillis = maxWaitMillis;
+        return this;
+    }
+
+    /**
+     * Performs the locking algorithm using the provided callbacks.
+     *
+     * @param factory         the producer which performs the action (creates an object etc.)
+     * @param correctnessTest the test which ensures that the action was successful without any collision
+     * @param cleanup         the callback which undoes the changes of <tt>factory</tt> if a collision was detected by
+     *                        <tt>correctnessTest</tt>
+     * @return the created object or an empty optional if the algorithm ran out of retries
+     * @throws Exception in case any of the provided callbacks threw an exception
+     */
+    public Optional<R> tryToCreate(Producer<R> factory, Processor<R, Boolean> correctnessTest, Callback<R> cleanup)
+            throws Exception {
+
+        int retries = numberOfRetries;
+        while (retries-- > 0) {
```

Comment:
This should be >= 0 or even < 0 ; depending if retries is really the number of "retries" or simply the number of "attempts".

Currently if i have retries = 5 there will only be 4 attempts as the first attempt already decreases the variable to 4 and on the 5. iteration 0 > 0 is false

Reply:
or do it like this:
```
        int retries = numberOfRetries;
        while (retries > 0) {
            retries--;
```

Reply:
while(retries-- > 0) ... and wile(retries > 0) ... retires---; is the same